src/cudamatrix/cu-kernels.cu

```diff
@@ -741,6 +741,16 @@ static void _set_bias_params(Real* v, const Real* a, Real param_1, Real param_2,
   }
 }
 
+template<typename Real, typename OtherReal>
+__global__
+static void _copy_from_vec(int n, const Real* x, int incx, OtherReal* y, int incy) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+   // if (blockIdx.y > 0) return;
+
+  if (i < n) {
+    y[1+(i-1)*incy] = (OtherReal) x[1+(i-1)*incx];
```

Comment:
Pegah, I think you are just confusing the issue.
The pointers that are passed into CUBLAS functions point to the first
elements of the array (they don't point to "just before the array").

If you want you could say that this is 13-based indexing, i.e. you could
say that the pointer points to element (13,13), that the indexes start from
13, and that to compute the indexes you have to use a complicated formula
involving subtracting 13 and then adding it again.

Best to just forget that 1-based indexing ever existed.  Your life will be
easier.


Dan

On Mon, May 23, 2016 at 3:02 PM, pegahgh <notifications@github.com> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/kaldi-asr/kaldi/pull/663#discussion_r64267409>:
>
> > @@ -741,6 +741,16 @@ static void _set_bias_params(Real* v, const Real* a, Real param_1, Real param_2,
> >    }
> >  }
> >
> > +template<typename Real, typename OtherReal>
> > +__global__
> > +static void _copy_from_vec(int n, const Real* x, int incx, OtherReal* y, int incy) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +   // if (blockIdx.y > 0) return;
> > +
> > +  if (i < n) {
> > +    y[1+(i-1)*incy] = (OtherReal) x[1+(i-1)*incx];
>
> cuBLAS API has both 0-based and 1-based indexing. For fortran code ported
> to C, 1-based indexing can be used for easier loop transfer. For C and C++
> code, we use 0-based indexing.
> Other point is that "i" is defined as blockIdx.x * blockDim.x +
> threadIdx.x and all blockIdx, threadIdx and blockDim ranges [0 ~ m) and i
> starts from 0 and you can not define (i-1)*incy even with 1-based
> indexing!!!!
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/663/files/86cab1989351bae78ec56730b008f7df8ec32834#r64267409>
>